### ISSUE TITLE
Building up search index in multi domain installation

### ISSUE THREAD

**Issue by bennyborn** [contao/core-bundle#1372]
Happened in version 4.5.2.

The customer got an installation with two page trees which are mapped to `example.de` and `example.com`.
Being logged in into the backend of `example.de` generating the search index for all .com sites will fail because of two reasons:

1. The `OPTIONS` request done via AJAX will fail because the .com domain is not allowed by the default CORS settings

```
Failed to load https://www.example.com/something/something.html: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://www.example.de' is therefore not allowed access.
```

2. If the above is fixed by adding e.g. `Header set Access-Control-Allow-Origin "*"` to the .htaccess of the installation the request will still fail with a 500 error. The logs revealed the following

```
[2018-02-14 19:28:47] request.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Type error: Return value of Contao\CoreBundle\Csrf\MemoryTokenStorage::getUsedTokens() must be of the type array, null returned" at /doesntmatter/vendor/contao/core-bundle/src/Csrf/MemoryTokenStorage.php line 95 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Type error: Return value of Contao\\CoreBundle\\Csrf\\MemoryTokenStorage::getUsedTokens() must be of the type array, null returned at /doesntmatter/vendor/contao/core-bundle/src/Csrf/MemoryTokenStorage.php:95)"} []
```

**Comment by fritzmg** [comment]
You need to set the domain in each document root, otherwise it will not work.

**Comment by bennyborn** [comment]
? Both domains are mapped to the same document root. It's one installation with two page trees.

**Comment by fritzmg** [comment]
Sorry, I meant the _website_ root settings within Contao. You need to set the domain in _both_ website root pages in your site structure.

**Comment by leofeyer** [comment]
@Toflar Isn't this why you have added the [WebsiteRootsConfigProvider](https://github.com/contao/core-bundle/blob/master/src/Cors/WebsiteRootsConfigProvider.php)?

**Comment by bennyborn** [comment]
@leofeyer Seems like that was the intended purpose but it's still missing the `OPTIONS` method

https://github.com/contao/core-bundle/blob/f2c9107899f11abdd5ec327cde35b80719deb973/src/Cors/WebsiteRootsConfigProvider.php#L62-L65

According to [MDN](https://developer.mozilla.org/en-US/docs/HTTP_access_control)

> Preflighted requests
> 
> Unlike simple requests (discussed above), "preflighted" requests first send an HTTP OPTIONS request header to the resource on the other domain, in order to determine whether the actual request is safe to send. Cross-site requests are preflighted like this since they may have implications to user data. In particular, a request is preflighted if:
> It uses methods other than GET or POST. Also, if POST is used to send request data with a Content-Type other than application/x-www-form-urlencoded, multipart/form-data, or text/plain, e.g. if the POST request sends an XML payload to the server using application/xml or text/xml, then the request is preflighted.
> 
> It sets custom headers in the request (e.g. the request uses a header such as X-PINGOTHER)

**Comment by fritzmg** [comment]
Which browser did you use?

**Comment by bennyborn** [comment]
Chrome 64.0.3282.140

**Comment by fritzmg** [comment]
> ```
> [2018-02-14 19:28:47] request.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Type error: Return value of Contao\CoreBundle\Csrf\MemoryTokenStorage::getUsedTokens() must be of the type array, null returned" at /doesntmatter/vendor/contao/core-bundle/src/Csrf/MemoryTokenStorage.php line 95 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Type error: Return value of Contao\\CoreBundle\\Csrf\\MemoryTokenStorage::getUsedTokens() must be of the type array, null returned at /doesntmatter/vendor/contao/core-bundle/src/Csrf/MemoryTokenStorage.php:95)"} []
> ```

This is fixed in Contao 4.5.4: https://github.com/contao/core-bundle/issues/1342

**Comment by fritzmg** [comment]
I cannot reproduce your problem (at least not in FireFox 58.0.2).

![screenshot-2018-2-15 maintenance - contao open source cms](https://user-images.githubusercontent.com/4970961/36248373-9025db68-1237-11e8-8a81-08ce242850b6.png)

I was logged in under `c45.dev`.

## OPTIONS request for `c45.local`:

### Request

```
Host: c45.local
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en,en-GB;q=0.8,de-AT;q=0.5,de;q=0.3
Accept-Encoding: gzip, deflate
Access-Control-Request-Method: GET
Access-Control-Request-Headers: x-requested-with
Origin: http://c45.dev
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
```
### Response
```
Date: Thu, 15 Feb 2018 09:00:27 GMT
Server: Apache/2.4.25 (Win32) OpenSSL/1.0.2j PHP/7.1.4
X-Powered-By: PHP/7.1.4
Cache-Control: no-cache, private
Access-Control-Allow-Methods: HEAD, GET
Access-Control-Allow-Headers: x-requested-with
Access-Control-Allow-Origin: http://c45.dev
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer-when-downgrade, strict-origin-when-cross-origin
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
Content-Length: 0
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8
```
## GET request for `c45.local`:

### Request
```
Host: c45.local
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0
Accept: text/javascript, text/html, application/xml, text/xml, */*
Accept-Language: en,en-GB;q=0.8,de-AT;q=0.5,de;q=0.3
Accept-Encoding: gzip, deflate
Referer: http://c45.dev/
X-Requested-With: XMLHttpRequest
Origin: http://c45.dev
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
```
### Response
```
Date: Thu, 15 Feb 2018 09:00:27 GMT
Server: Apache/2.4.25 (Win32) OpenSSL/1.0.2j PHP/7.1.4
Cache-Control: must-revalidate, no-cache, private
Vary: Contao-Page-Layout
Contao-Page-Layout: desktop
x-powered-by: PHP/7.1.4
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer-when-downgrade, strict-origin-when-cross-origin
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
Access-Control-Allow-Origin: http://c45.dev
Content-Length: 1750
Set-Cookie: csrf_contao_csrf_token=fhcVEBMKxWjrsH0S22x0A2yTS9fq_w3Rdzf4pEdOsl0; expires=Fri, 16-Feb-2018 09:00:27 GMT; Max-Age=86400; path=/; HttpOnly
Keep-Alive: timeout=5, max=99
Connection: Keep-Alive
Content-Type: text/html; charset=utf-8
 ```

**Comment by Toflar** [comment]
> @Toflar Isn't this why you have added the WebsiteRootsConfigProvider?

Yes, and it works for me. Maybe it's browser specific?

**Comment by leofeyer** [comment]
@bennyborn Can you please try again with Contao 4.5.4?

**Comment by bennyborn** [comment]
@leofeyer Didn't test it with 4.5.4 yet because I can't use the production system for that.
I'll try to setup a similar test environment with 4.5.4 tomorrow.
But I just tried using Firefox instead with the same result - OPTIONS method with a 500 response.

**Comment by Toflar** [comment]
The other error prevents the event listener from setting the correct headers which is why it doesn't work for you. Updating will solve the issue.

**Comment by bennyborn** [comment]
Confirmed. This is fixed in 4.5.4.
